### PR TITLE
Collaboration api endpoints and unit tests

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: boxsdk/object/collaboration.py [hunk 1]
```diff
@@ -57,3 +57,15 @@ def update_info(self, role=None, status=None):
...
+    def respond_to_pending(self, new_status):
+        """Accepts a pending collaboration
+
+        :param new_status:
+            The status of the collaboration invitation.
+        :type new_status:
+            `unicode`
+        """
+        data = {}
+        data['status'] = new_status
```
Comment (anthonywee) [inline boxsdk/object/collaboration.py]: could do:
> data = {'status': new_status}

File: boxsdk/object/file.py [hunk 2]
```diff
@@ -289,3 +291,34 @@ def get_shared_link_download_url(
...
+        :param marker:
+            The paging marker to start paging from.
+        :type marker:
+            `str` or None
+        :param fields:
+            List of fields to request.
+        :type fields:
+            `Iterable` of `unicode`
+        :returns:
+            An iterator of the entries in the legal hold policy
```
Comment (anthonywee) [inline boxsdk/object/file.py]: legal hold policy?

File: boxsdk/object/file.py [hunk 3]
```diff
@@ -289,3 +291,34 @@ def get_shared_link_download_url(
             password=password,
         )
         return item.shared_link['download_url']
+
+
+    def collaborators(self, limit=None, marker=None, fields=None):
+        """
+        Get the entries in the collaborators using limit-offset paging.
```
Comment (anthonywee) [inline boxsdk/object/file.py]: marker based

File: boxsdk/object/file.py [hunk 4]
```diff
@@ -289,3 +291,34 @@ def get_shared_link_download_url(
...
+        :rtype:
+            :class:`BoxObjectCollection`
+        """
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url='{0}/files/{1}/collaborations'.format(API.BASE_API_URL, self.object_id),
+            limit=limit,
+            marker=marker,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/file.py]: trailing comma

File: boxsdk/object/folder.py [hunk 5]
```diff
@@ -382,3 +383,34 @@ def delete(self, recursive=True, etag=None):
...
+        :rtype:
+            :class:`BoxObjectCollection`
+        """
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url='{0}/folders/{1}/collaborations'.format(API.BASE_API_URL, self.object_id),
+            limit=limit,
+            marker=marker,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/folder.py]: trailing comma

File: boxsdk/object/item.py [hunk 6]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+        :type role:
+            `unicode`
+        :param accessible_by:
+            An object containing the collaborator.
+        :type etag:
+            `object`
+        :param accessible_by:
+            Indicates whether the user can view the path of the folder collaborated into.
+        :type can_view_path:
+            `boolean`
```
Comment (anthonywee) [inline boxsdk/object/item.py]: "`bool`"

File: boxsdk/object/item.py [hunk 7]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+        :param notify:
+            Determines if the collaborator should receive a notification for the collaboration.
+        :type notify:
+            `boolean`
```
Comment (anthonywee) [inline boxsdk/object/item.py]: `bool`

File: boxsdk/object/item.py [hunk 8]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
         """
         headers = {'If-Match': etag} if etag is not None else None
         return super(Item, self).delete(params, headers)
+
+
+    def collaborate(self, role, accessible_by, can_view_path=None, notify=None, fields=None):
```
Comment (anthonywee) [inline boxsdk/object/item.py]: is can_view_path used?

File: boxsdk/object/item.py [hunk 9]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+        """Collaborate user or group onto a Box item.
+
+        :param role:
+            The permission level to grant the collaborator.
+        :type role:
+            `unicode`
+        :param accessible_by:
+            An object containing the collaborator.
+        :type etag:
+            `object`
```
Comment (anthonywee) [inline boxsdk/object/item.py]: Can this be more specific than `object`?

File: boxsdk/object/item.py [hunk 10]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+            `boolean`
+        :param fields:
+            List of fields to request.
+        :type fields:
+            `Iterable` of `unicode`
+        """
+        url = self._session.get_url('collaborations')
+        body = {
+            'item': {
+                'type': self.objecttype,
```
Comment (anthonywee) [inline boxsdk/object/item.py]: object_type?

File: boxsdk/object/base_object.py [hunk 11]
```diff
@@ -219,3 +219,11 @@ def clone(self, session=None):
             self._object_id,
             self._response_object,
         )
+
+    @property
+    def object_type(self):
```
Comment (anthonywee) [inline boxsdk/object/base_object.py]: Is this needed, since the BaseAPIJSONObject already defines it?

File: boxsdk/object/item.py [hunk 12]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+        :rtype:
+            :class:`BoxObjectCollection`
+        """
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url=self.get_url('collaborations'),
+            limit=500,
+            marker=None,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/item.py]: trailing comma

File: boxsdk/object/item.py [hunk 13]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+            params['notify'] = ','.join(fields)
+        response = self._session.post(url, data=json.dumps(body), params=params).json()
+        return Translator().translate(response['type'])(
+            self._session,
+            response['id'],
+            response,
+        )
+
+
+    def collaborate_with_login(self, role, login, can_view_path=None, notify=None, fields=None):
```
Comment (anthonywee) [inline boxsdk/object/item.py]: are we using can_view_path?

File: boxsdk/object/item.py [hunk 14]
```diff
@@ -322,3 +325,161 @@ def delete(self, params=None, etag=None):
...
+        if fields:
+            additional_params['fields'] = ','.join(fields)
+        return LimitOffsetBasedObjectCollection(
+            session=self._session,
+            url='{0}/collaborations'.format(API.BASE_API_URL),
+            additional_params=additional_params,
+            limit=limit,
+            offset=offset,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/item.py]: trailing comma

File: test/unit/object/test_item.py [hunk 15]
```diff
@@ -127,3 +128,117 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+
+
+def test_collaborate(test_item_and_response, test_group, mock_box_session):
+    # pylint:disable=redefined-outer-name, protected-access
+    test_item, _ = test_item_and_response
+    expected_url = test_item.get_url('collaborations')
+    expected_data = {
+        'item': {
+            'type': test_item.object_type,
+            'id': test_item.object_id
```
Comment (anthonywee) [inline test/unit/object/test_item.py]: trailing comma

File: test/unit/object/test_item.py [hunk 16]
```diff
@@ -127,3 +128,117 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+        },
+        'accessible_by': {
+            'type': test_group.object_type,
+            'id': test_group.object_id
```
Comment (anthonywee) [inline test/unit/object/test_item.py]: trailing comma

File: test/unit/object/test_item.py [hunk 17]
```diff
@@ -127,3 +128,117 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+        },
+        'role': 'editor'
```
Comment (anthonywee) [inline test/unit/object/test_item.py]: trailing comma

File: test/unit/object/test_item.py [hunk 18]
```diff
@@ -127,3 +128,117 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+    }
+    mock_collaboration = {
+        'type': 'collaboration',
+        'id': '1234',
+        'created_by': {
+            'type': 'user',
+            'id': '1111'
```
Comment (anthonywee) [inline test/unit/object/test_item.py]: trailing comma

File: test/unit/object/test_item.py [hunk 19]
```diff
@@ -127,3 +128,117 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+    assert collaboration['type'] == mock_collaboration['type']
+
+def test_collaborate_with_login(test_item_and_response, mock_box_session):
+    # pylint:disable=redefined-outer-name, protected-access
+    test_item, _ = test_item_and_response
+    expected_url = mock_box_session.get_url('collaborations')
+    expected_data = {
+        'item': {
+            'type': test_item.object_type,
+            'id': test_item.object_id
```
Comment (anthonywee) [inline test/unit/object/test_item.py]: trailing commas here and below

**Pull request closed without merging**